### PR TITLE
PARQUET-229 Add a strict thrift projection API with backwards compat support

### PR DESCRIPTION
**Pull request by isnotinvain**
Currently, the thrift projection API accepts strings in a very general glob format that supports not only wildcards like `*` and `?` and expansions (`{x,y,z}`) but also character classes `[abc]`, and negation.
Because of this flexibility, it's hard to give users good error reporting, for example letting them know that when they requested columns `foo.bar.{a,b,c}` there is actually no such column `foo.bar.c`.

This PR introduces a new syntax that supports a more restrictive form of glob syntax and enforces that all **expansions** of a glob match a column, not just that all globs match a column. The new syntax is very simple and only has four special characters: `{`,`}`,`,`, and `*`

It supports glob expansion, for example:
`home.{phone,address}` or `org.apache{-incubator,}.foo`

And the wildcard `*` which is treated the same way as java regex treats `(.*)`, for example:
`home.*` or `org.apache*.foo`

In the new syntax glob paths mean "keep all the child fields of the field matched by this glob", just like variable access would work in a programming language. For example: `x.y.z` means keep field `z` and all of its children (if any). So it's not necessary to do `x.y.z.*`. However, `x.y.z` would not keep field `x.y.zoo`. If that was desired, then `x.y.z*` could be used instead.

Setting `"parquet.thrift.column.filter"` will result in the same behavior that it does currently in master, though a deprecation warning will be logged. The classes that implement the current behavior have been marked as deprecated, and using this will log a warning.

Setting `"parquet.thrift.column.projection.globs"` will instead use this new syntax, and entry points in the various Builder's in the codebase is added as well.

This PR does a little bit of cleanup as well, moving some shared methods to a `Strings` class and simplifying some of the class hierarchy in `ThriftSchemaConverterVisitor`. There are a few `// TODO Why?` added as well that I wanted to ask about.

### CODE REVIEW

**Comment by isnotinvain** [comment]
cc @tsdeng @julienledem

File: parquet-cascading/src/main/java/parquet/cascading/ParquetValueScheme.java [hunk 1]
```diff
@@ -105,9 +119,17 @@ public ParquetValueScheme(Config<T> config) {
     this.config = config;
   }
 
+  // TODO: Why is there thrift references in this file?
```
Comment (isnotinvain) [inline parquet-cascading/src/main/java/parquet/cascading/ParquetValueScheme.java]: just wanted to call this question out

File: parquet-common/src/main/java/parquet/glob/GlobParser.java [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+    int anchor = 0; // first un-parsed character position
+
+    for (int i = 0; i < pattern.length(); i++) {
+      char c = pattern.charAt(i);
+
+      switch (c) {
+        case ',':
+          if (unmatchedBraces == 0) {
+            // commas not allowed in the top level expression
+            // TODO: maybe turn this check off?
```
Comment (isnotinvain) [inline parquet-common/src/main/java/parquet/glob/GlobParser.java]: what do you think?

File: parquet-thrift/src/main/java/parquet/thrift/ThriftSchemaConvertVisitor.java [hunk 3]
```diff
@@ -49,25 +50,21 @@
...
-  }
-
-  FieldProjectionFilter fieldProjectionFilter;
-  Type currentType;
-  FieldsPath currentFieldPath = new FieldsPath();
-  Type.Repetition currentRepetition = Type.Repetition.REPEATED;//MessageType is repeated GroupType
-  String currentName = "ParquetSchema";
+  private Type currentType;
+  private Type.Repetition currentRepetition = Type.Repetition.REPEATED; // MessageType is repeated GroupType
+  private String currentName = "ParquetSchema"; // TODO: why?
```
Comment (isnotinvain) [inline parquet-thrift/src/main/java/parquet/thrift/ThriftSchemaConvertVisitor.java]: just want to call out this question

File: parquet-thrift/src/main/java/parquet/thrift/ThriftSchemaConvertVisitor.java [hunk 4]
```diff
@@ -90,7 +87,7 @@ public void visit(ThriftType.MapType mapType) {
     //=========handle value
     currentFieldPath.push(mapValueField);
     currentName = "value";
-    currentRepetition = OPTIONAL;
+    currentRepetition = OPTIONAL; // TODO: why?
```
Comment (isnotinvain) [inline parquet-thrift/src/main/java/parquet/thrift/ThriftSchemaConvertVisitor.java]: Why are map value optional? If they get projected out, we replace them with null?

File: parquet-thrift/src/main/java/parquet/thrift/ThriftSchemaConvertVisitor.java [hunk 5]
```diff
@@ -100,6 +97,7 @@ public void visit(ThriftType.MapType mapType) {
       return;
     }
 
+    // TODO: this can never happen
     if (keyType == null && valueType != null)
```
Comment (isnotinvain) [inline parquet-thrift/src/main/java/parquet/thrift/ThriftSchemaConvertVisitor.java]: Intellij points out that this if statement is unreachable -- is this a bug?

**Comment by isnotinvain** [comment]
Tests running here: https://travis-ci.org/isnotinvain/incubator-parquet-mr/builds/55340816

File: parquet-cascading/src/main/java/parquet/cascading/ParquetValueScheme.java (hunk 1 above)
Comment (julienledem) [inline parquet-cascading/src/main/java/parquet/cascading/ParquetValueScheme.java]: Filters are defined in terms of the thrift schema and T has to be either a Thrift of Scrooge generated class.
Possibly this should be refactored or renamed.

File: parquet-thrift/src/main/java/parquet/thrift/ThriftSchemaConvertVisitor.java (hunk 4 above)
Comment (tsdeng) [inline parquet-thrift/src/main/java/parquet/thrift/ThriftSchemaConvertVisitor.java]: yes, the value of a map can be projected out

Comment (isnotinvain) [inline parquet-thrift/src/main/java/parquet/thrift/ThriftSchemaConvertVisitor.java]: can parts of the key be projected out? That wouldn't make any sense right?

Comment (isnotinvain) [inline parquet-thrift/src/main/java/parquet/thrift/ThriftSchemaConvertVisitor.java]: I should double check if we enforce that you can't project fields nested under a key

File: parquet-cascading/src/main/java/parquet/cascading/ParquetValueScheme.java (hunk 1 above)
Comment (isnotinvain) [inline parquet-cascading/src/main/java/parquet/cascading/ParquetValueScheme.java]: ok

**Comment by isnotinvain** [comment]
tests passed: https://travis-ci.org/isnotinvain/incubator-parquet-mr/builds/57598699

**Comment by tsdeng** [comment]
could you also update this doc with the new API? https://github.com/apache/incubator-parquet-mr/blob/master/parquet_cascading.md
It could serve as a tutorial for users to use it

**Comment by tsdeng** [comment]
minor comment, otherwise LGTM

File: parquet-thrift/src/main/java/parquet/thrift/ThriftSchemaConvertVisitor.java (hunk 3 above)
Comment (isnotinvain) [inline parquet-thrift/src/main/java/parquet/thrift/ThriftSchemaConvertVisitor.java]: revisit this later

**Pull request closed without merging**